lib/cpp/scalgoproto.hh

```diff
@@ -771,8 +777,70 @@ struct MetaMagic<ListOut<A>> {
 	using t = ListTag;
 };
 
+class WriterBacking {
+public:
+	virtual char * setCapacity(size_t newCapacity) = 0;
+};
+
+class FileWriterBacking : public WriterBacking {
+private:
+	int fd = -1;
+	void * addr = nullptr;
+	size_t capacity = 0;
+
+	static constexpr size_t PAGE_SIZE = 4096;
+
+public:
+	int open(const char * path) {
+		assert(::sysconf(_SC_PAGE_SIZE) == PAGE_SIZE);
+		if (addr != nullptr) return -1;
+		fd = ::open(path, O_TRUNC | O_CREAT | O_RDWR, 0666);
+		if (fd == -1) return -1;
+		if (::ftruncate(fd, PAGE_SIZE)) {
+			::close(fd);
```

Comment:
Use are using errno after a call to open failes, so perhaps we should restore errno after close, as we would want the errno of truncate and not close?

Reply:
Thanks - fixed.